### PR TITLE
:running: e2e k8s conformance

### PR DESCRIPTION
**Pull request by bhargavmad**
:running:
**What this PR does / why we need it**:
Runs k8s conformance tests on workload cluster as part of e2e testing
 - Uses sonobuoy library to achieve it
 - Outputs a result tarball

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -87,6 +87,11 @@ test-e2e: ## Run e2e tests
 	PULL_POLICY=IfNotPresent $(MAKE) docker-build
 	go test -v -tags=e2e -timeout=1h ./test/e2e/... -args -ginkgo.v --managerImage $(CONTROLLER_IMG)-$(ARCH):$(TAG)
```
Comment (detiber) [inline Makefile]: Do we also need to add a focus here as well? Otherwise wouldn't this make target also cause the conformance tests to run?

File: test/e2e/aws_test.go [hunk 2]
```diff
@@ -107,7 +107,10 @@ var _ = Describe("functional tests", func() {
 			machineName = cpMachinePrefix + "-2"
 			createAdditionalControlPlaneMachine(namespace, clusterName, machineName, awsMachineName, bootstrapConfigName)
 
-			// TODO: Deploy a MachineDeployment
+			By("Deploying a MachineDeployment")
+			machineDeploymentName := clusterName + "-worker"
+			createMachineDeployment(namespace, clusterName, machineDeploymentName, 3)
+
```
Comment (detiber) [inline test/e2e/aws_test.go]: Are these changes needed? It looks like the required changes are in the helper methods rather than here. 

This will also need to be reconciled with https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1247 after it is merged.

File: test/e2e/aws_test.go [hunk 3]
```diff
@@ -406,7 +517,7 @@ func makeAWSMachine(namespace, name string) {
 			Namespace: namespace,
 		},
 		Spec: infrav1.AWSMachineSpec{
-			InstanceType:       "t3.large",
+			InstanceType:       "m5.large",
```
Comment (detiber) [inline test/e2e/aws_test.go]: This should probably be exposed as a parameter for `makeAWSMachine` to limit costs for running the tests contained in `aws_test.go` compared to `conformance_test.go`

File: Makefile (hunk 1 above)
Comment (bhargavmad) [reply Makefile]: Oh yes. i missed that. Thanks.

File: go.mod [hunk 4]
```diff
@@ -24,3 +26,5 @@ require (
 	sigs.k8s.io/controller-runtime v0.3.0
 	sigs.k8s.io/yaml v1.1.0
 )
+
+replace k8s.io/client-go => k8s.io/client-go v0.0.0-20190918200256-06eb1244587a
```
[the remark this answers is not in the document]
Comment (bhargavmad) [reply go.mod]: As I understand, sonobuoy has this dependency on a particular version of k8s/client-go and getting a newer version of client-go is breaking it. I hate to do this as well but i couldnt find a better way.

File: test/e2e/aws_test.go (hunk 2 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: Yes. I should probably wait for #1247

File: test/e2e/aws_test.go (hunk 3 above)
Comment (bhargavmad) [reply test/e2e/aws_test.go]: Sounds good. Will do.

File: test/e2e/conformance_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	eg := &errgroup.Group{}
+	eg.Go(func() error { return <-ec })
+	eg.Go(func() error {
+		filesCreated, err := client.UntarAll(reader, ".", "")
+		if err != nil {
+			return errors.Wrap(err, "couldn't untar sonobuoy results")
+		}
+		for _, name := range filesCreated {
+			fmt.Fprintf(GinkgoWriter, "Find conformance test results at: test/e2e/%s\n", name)
```
Comment (johnSchnake) [inline test/e2e/conformance_test.go]: Unless I'm misreading this, I believe that the `filesCreated` is the tarball itself and so the path "test/e2e/<name>" isn't right. 

E.g. when you run `sonobuoy retrieve` it prints the filenames of the tarball it downloaded. Maybe this was misleading because of the `UntarAll` but it is untarring the response from the server which in most cases results in a single tarball name in `filesCreated` but (in theory) could be more tarballs.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply go.mod]: If we can't fix this here, I propose we make `test/e2e` its own go module with the `replace` statement there.

File: test/e2e/sonobuoy-config-tmpl.yaml [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      command:
+      - /run_e2e.sh
+      env:
+      - name: E2E_FOCUS
+        value: \[Conformance\]
+      - name: E2E_SKIP
+      - name: E2E_PARALLEL
+        value: "1"
+      - name: E2E_USE_GO_RUNNER
+        value: "true"
```
Comment (johnSchnake) [inline test/e2e/sonobuoy-config-tmpl.yaml]: As we spoke about on slack; this will only be useful if you have a v1.16+ cluster since the gorunner didnt exist before that.

File: test/e2e/sonobuoy-config-tmpl.yaml [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
...ents","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations","namespaces","networkpolicies","nodes","persistentvolumeclaims","persistentvolumes","poddisruptionbudgets","pods","podlogs","podsecuritypolicies","podtemplates","priorityclasses","replicasets","replicationcontrollers","resourcequotas","rolebindings","roles","servergroups","serverversion","serviceaccounts","services","statefulsets","storageclasses","validatingwebhookconfigurations","volumeattachments"],"Filters":{"Namespaces":".*","LabelSelector":""},"Limits":{"PodLogs":{"Namespaces":"","SonobuoyNamespace":true,"FieldSelectors":[],"LabelSelector":"","Previous":false,"SinceSeconds":null,"SinceTime":null,"Timestamps":false,"TailLines":null,"LimitBytes":null,"LimitSize":"","LimitTime":""}},"QPS":30,"Burst":50,"Server":{"bindaddress":"0.0.0.0","bindport":8080,"advertiseaddress":"","timeoutseconds":10800},"Plugins":[{"name":"e2e"},{"name":"systemd-logs"}],"PluginSearchPath":["./plugins.d","/etc/sonobuoy/plugins.d","~/sonobuoy/plugins.d"],"Namespace":"sonobuoy","WorkerImage":"sonobuoy/sonobuoy:{{.SonobuoyVersion}}","ImagePullPolicy":"IfNotPresent","ImagePullSecrets":"","ProgressUpdatesPort":"8099"}
```
[the remark this answers is not in the document]
Comment (bhargavmad) [reply test/e2e/sonobuoy-config-tmpl.yaml]: Yes. Even I didnt like the way it is. But didnt know a better way. Any suggestion?

[the remark this answers is not in the document]
Comment (johnSchnake) [reply test/e2e/sonobuoy-config-tmpl.yaml]: I'm 95% sure that if you just delete the UUID then it will be dynamically generated on the server when loaded.

**Comment by ncdc** [comment]
I want to avoid needing to use a `replace` directive in CAPA's main `go.mod` file... we could either make `test/e2e` a separate go module, or switch to running sonobuoy either as a binary, or as a pod.

also
/cc @chuckha

**Comment by johnSchnake** [comment]
In general the approach of executing the run using Sonobuoy as a library is just fine; I agree with the approach of using the `GenFile` so that you can look at/tweak the raw YAML that will be used and not even worry about all the `GenConfig`->YAML translation.

I think that doing it this way (using Sonobuoy as a library) will be easier to maintain long term:
 - No need to build/publish/maintain the extra wrapper and logic
 - By avoiding using the CLI in a wrapper you really limit the amount of Sonobuoy you're using which means less surface area for bugs/etc. You'll only want/need to update as core server features are updated if you want them.
 - By avoiding use of the CLI you wont need to worry about us maintaining that CLI level compatibility; we may break CLI compatibility without too much thought but are more careful when it comes to the config object itself.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (detiber) [reply go.mod]: :grimacing: I think I would rather do the replace in the main `go.mod` file if we need to override it for sonobuoy, otherwise we risk shipping something different than we are testing.

File: test/e2e/sonobuoy-config-tmpl.yaml [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    sonobuoy-config:
+      driver: Job
+      plugin-name: e2e
+      result-format: junit
+    spec:
+      command:
+      - /run_e2e.sh
+      env:
+      - name: E2E_FOCUS
+        value: \[Conformance\]
```
[the remark this answers is not in the document]
Comment (detiber) [reply test/e2e/sonobuoy-config-tmpl.yaml]: Disruptive tests should be fine for our use case, since we are looking to test conformance specifically here and these clusters will not be used for other purposes.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (detiber) [reply go.mod]: I take that back, as long as we can isolate the system under test, ensuring that the controller manager that is built is using the main `go.mod` file and only the test suite/utilities are using a different version of client-go, then this approach should be fine. It might also help to keep other testing dependencies from creeping into the main `go.mod` file as well.

**Comment by bhargavmad** [comment]
@chuckha can you please have a look at this PR

**Comment by chuckha** [comment]
yep, sorry for the delay, looking now

File: test/e2e/conformance_test.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfigData)
+	if err != nil {
+		return errors.Wrap(err, "couldn't get rest config from kubeconfig")
+	}
+
+	sonobuoyKubeCli, err := sonodynamic.NewAPIHelperFromRESTConfig(restConfig)
+	if err != nil {
+		return err
+		return errors.Wrap(err, "couldn't get rest config from kubeconfig")
```
Comment (chuckha) [inline test/e2e/conformance_test.go]: unreachable code

File: test/e2e/conformance_test.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+	if err := sonobuoyClient.Run(runConfig); err != nil {
+		return errors.Wrap(err, "error attempting to run sonobuoy")
+	}
+	reader, ec, err := sonobuoyClient.RetrieveResults(&client.RetrieveConfig{Namespace: "sonobuoy"})
+	if err != nil {
+		return errors.Wrap(err, "couldn't retrieve sonobuoy results")
+	}
+	eg := &errgroup.Group{}
+	eg.Go(func() error { return <-ec })
+	eg.Go(func() error {
```
Comment (chuckha) [inline test/e2e/conformance_test.go]: Are we doing anything with the results or do you think it's enough that if we're able to retrieve the results then they must have passed?

**Comment by chuckha** [comment]
we may also be able to move this into the e2e test framework in capi at some point and provide conformance as an importable test

File: test/e2e/conformance_test.go (hunk 10 above)
Comment (detiber) [reply test/e2e/conformance_test.go]: It should likely been written out to the artifact path, similar to what we are currently doing in the e2e suite: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/test/e2e/e2e_suite_test.go#L70-L77

As long as the file is named junit.*.xml it will be picked up by test-grid.

Comment (chuckha) [reply test/e2e/conformance_test.go]: Oh, I see 👍 

consider adding a comment here

**Comment by bhargavmad** [comment]
@chuckha @detiber I have addressed all the comments. Please have a look.
Also, i am not very sure of the rebase procedure and i am scared i might mess up things. @detiber can you please help me out.

File: test/e2e/e2e_suite_test.go [hunk 11]
```diff
@@ -95,13 +95,14 @@ var (
...
-	suiteTmpDir string
-	region      string
+	kindCluster  kind.Cluster
+	kindClient   crclient.Client
+	sess         client.ConfigProvider
+	accountID    string
+	accessKey    *iam.AccessKey
+	suiteTmpDir  string
+	region       string
+	artifactPath string
```
Comment (bhargavmad) [inline test/e2e/e2e_suite_test.go]: I see that `artifactPath` is not part of the `release-0.4` branch. So this doesnt do anything now. But when merged with master, it should place the conformance results in the `artifactPath`

**Comment by dims** [comment]
may i suggest avoiding `conformance`? it has a specific connotation in our community.  `validation` may be a better term.

**Comment by detiber** [comment]
@dims this does indeed run k8s conformance using sonobuoy

**Comment by dims** [comment]
ah excellent, sorry about the noise :)

**Comment by bhargavmad** [comment]
@detiber can you push the clean branch onto this and merge if no changes required. Thanks.
